pasha/functor.py

```diff
@@ -0,0 +1,280 @@
+        """
+
+        self.sequence = sequence
+
+    @classmethod
+    def wrap(cls, value):
+        if isinstance(value, Sequence):
+            return cls(value)
+
+    def split(self, num_workers):
+        return gen_split_slices(len(self.sequence), n_parts=num_workers)
+
+    def iterate(self, share):
+        yield from zip(range(*share.indices(len(self.sequence))),
+                       self.sequence[share])
+
+
+class NdarrayFunctor(SequenceFunctor):
+    """Functor wrapping an numpy.ndarray.
+
+    This functor extends SequenceFunctor to allow iterating over
+    specific axes. While it should work for any array_like object,
+    specifying an explicit axis may cause conversion to an ndarray.
+    """
+
+    def __init__(self, array, axis=None):
+        """Initialize an ndarray functor.
+
+        Args:
+            array (numpy.ndarray): Array to map over.
+            axis (int, optional): Axis to map over, first axis by
+                default or if None.
+        """
+
+        self.sequence = np.swapaxes(array, 0, axis) \
+            if axis is not None else array
+
+    @classmethod
+    def wrap(cls, value):
+        if isinstance(value, np.ndarray):
+            return cls(value)
+
+
+class DataArrayFunctor(NdarrayFunctor):
+    """Functor wrapping an xarray.DataArray.
+
+    This functor extends NdarrayFunctor for compatbility with xarray's
+    DataArray type, in particular dimension labels and coordinates.
+    """
+
+    def __init__(self, array, dim=None):
+        """Initialize a DataArray functor.
+
+        Args:
+            array (xarray.DataArray): Labeled array to map over.
+            dim (str, optional): Dimension to map over, first dimension
+                by default or if None.
+        """
+
+        self.sequence = array.transpose(dim, ...) \
+            if dim is not None else array
+
+    @classmethod
+    def wrap(cls, value):
+        if 'xarray' not in sys.modules:
+            # If xarray has not been loaded yet, we can safely assume
+            # this to not be a DataArray.
+            return
+
+        import xarray as xr
+
+        if isinstance(value, xr.DataArray):
+            return cls(value)
+
+
+# Ideas for wrapping functors: xarray.Dataset, pandas
+
+
+class ExtraDataFunctor(Functor):
+    """Functor for EXtra-data objects.
+
+    This functor wraps an EXtra-data DataCollection or KeyData and
+    performs the map operation over its trains. The kernel is passed the
+    current train's index in the collection, the train ID and the data
+    mapping (for DataCollection) or data entry (for KeyData).
+    """
+
+    def __init__(self, obj):
+        self.obj = obj
+        self.n_trains = len(self.obj.train_ids)
+
+    @classmethod
+    def wrap(cls, value):
+        if 'extra_data' not in sys.modules:
+            # Same assumption as in DataArrayFunctor.
+            return
+
+        import extra_data as xd
+
+        if isinstance(value, (xd.DataCollection, xd.keydata.KeyData)):
```

Comment:
From 1.4, KeyData will also be available as a top-level name `xd.KeyData`. You can't rely on that yet, just a heads up that it should one day.